### ISSUE TITLE
Question about job distribution

### ISSUE THREAD

**Issue by atis--** [grantcarthew/node-rethinkdb-job-queue#62]
Hi!

Wow, this is an impressive and well-documented project! Browsing through the
docs/source I can't find an answer to the following question however.

Let's say I have multiple subscriber processes/machines, all waiting for new
jobs on the same high volume queue. I'm assuming the subscribers have a
changefeed that listens for new jobs being inserted into the queue table? But
then how is it ensured that all subscribers do not race to claim the same jobs?

Basically I want the jobs distributed in such a way that each subscriber gets
their own jobs, without the others receiving the same traffic. Is that how
things will happen, and if so, then how is that achieved?

Thanks in advance.

**Comment by grantcarthew** [comment]
Hi @atis-- and thanks for the kind words.

The answer to this question is based on two features, one in the`rethinkdb-job-queue` module and one in RethinkDB.

Firstly, the `rethinkdb-job-queue` has a [concurrency](https://github.com/grantcarthew/node-rethinkdb-job-queue/wiki/Queue-Options#queue-concurrency-option) option that restricts how many jobs each worker node will retrieve.

Secondly, the queries to retrieve the jobs from the database are atomic.

You're right about the changefeed. Each worker node receives changes letting them know there are new jobs on the queue. Which worker node gets the first "concurrency" number of jobs would be based on network latency and other factors. Once the first worker node retrieves the jobs they are taken off the queue atomically. The next worker will get the next "concurrency" number of jobs and so on until all the jobs are gone or all the workers are at "concurrency".

Hope this helps answer your question.

**Comment by atis--** [comment]
My concern is not so much the atomicity/concurrency (which is clearly handled
correctly), but the fact that the changefeed in each worker is getting ALL the
changes that occur on the queue table (which includes the complete job entry
data for all inserts and status updates).

So if I have 5 workers, and insert one job, all 5 of them are going to get the
data even though only one of them ends up handling the job. So adding a worker
would actually increase the network & (changefeed) processing load for the database and all
workers in a way that seems... wasteful?

**Comment by grantcarthew** [comment]
Your understanding is correct @atis-- 

When I first designed `rethinkdb-job-queue` I decided to make each worker node watch for changes on the whole queue backing table and manage the change logic locally. For reference this is all happening in the [queue-change.js](https://github.com/grantcarthew/node-rethinkdb-job-queue/blob/master/src/queue-change.js) file.

This could be revisited however at the time it seemed like an easier way of managing a complex state system.

One of the features I have not tried to achieve with `rethinkdb-job-queue` is performance. I have not done any performance testing or comparison. My main goal was to remove a dependency on another database system (redis) for a distributed job queue.

If you are trying to get the best performance from a distributed worker queue I would suggest either RabbitMQ or some other queue system.

With that in mind, I have not testing the performance so it may be OK? I don't know.

If you are keen, please feel free to try moving the logic in the queue-change.js file into individual changefeed filters. Note that the queue-db.js file is where the current change feed is being registered.

Without thinking about it too heavily, the changefeed registration logic would need to be customized based on the options supplied to the Queue object. It would be a big job to change it.

**Comment by atis--** [comment]
Yeah, I sympathize with the idea of removing extra dependencies (redis, rabbit),
which is part of the allure of RethinkDB's changefeeds.

It does feel like there is room for a library that pushes the task of pubsub
job distribution onto filtered/indexed changefeeds. I don't know if
`rethinkdb-job-queue` (which already does complex job state management) is the
right place for doing that or not. As you mention, surely the logic would only
get more complex still.

Anyway, thanks for your quick and thorough response. Certainly some food for
thought.
